File: LeetCode/C/20-stack-有效的括号.cpp
/*
 * @lc app=leetcode.cn id=20 lang=cpp
 *
 * [20] 有效的括号
 */
#include "main.h"
// @lc code=start
class Solution {
  public:
    bool isValid(string s) {
        if (s.empty())
            return true;
        stack<char> st;
        for (char c : s) {
            if (c == '(')
                st.push(')');
            else if (c == '[')
                st.push(']');
            else if (c == '{')
                st.push('}');
            else if (c == ')' || c == ']' || c == '}') {
                if (st.empty() || st.top() != c)
                    return false;
                st.pop();
            } else
                return false;
        }
        return st.empty();
    }
};
// @lc code=end

File: LeetCode/Python/20-stack-有效的括号.py
#
# @lc app=leetcode.cn id=20 lang=python3
#
# [20] 有效的括号
#


# @lc code=start
class Solution:
    def isValid(self, s: str) -> bool:
        m = {")": "(", "]": "[", "}": "{"}
        stack = []
        for c in s:
            if c in m:
                if len(stack) == 0 or stack[-1] != m[c]:
                    return False
                stack.pop()
            else:
                stack.append(c)
        return len(stack) == 0


# @lc code=end
